discojs/src/client/federated/federated_client.ts

```diff
@@ -0,0 +1,247 @@
+    // Prepare the result promise for the incoming round
+    this.aggregationResult = new Promise((resolve) => this.aggregator.once('aggregation', resolve))
+    return Promise.resolve();
+  }
+
+  /**
+   * 
+   * @param weights Local weights sent to the server at the end of the local training round
+   * @param _round The trainer's aggregation round, which can be different from 
+   * the server's round if the participant joined late
+   * @returns the new global weights sent by the server
+   */
+  override async onRoundEndCommunication(weights: WeightsContainer): Promise<WeightsContainer> {
+    // NB: For now, we suppose a fully-federated setting.
+
+    if (this.aggregationResult === undefined) {
+      throw new Error("local aggregation result was not set");
+    }
+
+    // First we check if we are waiting for more participants before sending our weight update
+    if (this.#waitingForMoreParticipants) {
+      if (this.#promiseForMoreParticipants === undefined) throw new Error("Promise should not be undefined")
+        
+      // wait for the promise to resolve, which takes as long as it takes for new participants to join
+      debug("Awaiting the promise for more participants")
+      this.logger.setStatus("Waiting for more participants")
+      await this.#promiseForMoreParticipants 
+      debug("Promise resolved")
+    }
+    this.logger.setStatus("Updating the model with other participants' models")
+
+    // Send our local contribution to the server
+    // and receive the most recent weights as an answer to our contribution
+    const payload: WeightsContainer = this.aggregator.makePayloads(weights).first()
+    const msg: messages.SendPayload = {
+      type: type.SendPayload,
+      payload: await serialization.weights.encode(payload),
+      round: this.aggregator.round,
+    };
+
+    // Need to await the resulting global model right after sending our local contribution
+    // to make sure we don't miss it
+    this.server.send(msg);
+    const serverResult = await this.receiveServerResult();
+
+    if (
+      serverResult !== undefined &&
+      this.aggregator.add(FederatedClient.SERVER_NODE_ID, serverResult, this.aggregator.round)
+    ) {
+      // Regular case: the server sends us its aggregation result which will serve our
+      // own aggregation result.
+    } else {
+      // Unexpected case: for some reason, the server result is stale.
+      debug(`[${this.ownId}] server result for round ${this.aggregator.round} is undefined or stale`);
+      // We proceed to the next round without its result.
+      this.aggregator.nextRound();
+    }
+    return await this.aggregationResult
+  }
+
+  /**
+   * Wait for the server to reply with the most recent aggregated weights
+   * after having sent out local weight update to the federated server
+   */
+  private async receiveServerResult(): Promise<WeightsContainer | undefined> {
+    // Try receiving the latest update multiple times in case we receive outdated updates
```

Comment:
It's to handle unreliable network events like message swaps or an old update arriving late. But I guess an unreliable network would break many other things in disco, I'm mostly worried about only waiting for a single message because if it's not the right one then the participant can't recover for this round.

Reply:
I've implemented it in a loop but the idea simply is "wait for the expected update for a given amount of time and discard any unexpected message". But because we are manually "waiting" for server messages, "discarding unexpected messages" translates into a loop of manually waiting.

Reply:
> I'm mostly worried about only waiting for a single message because if it's not the right one then the participant can't recover for this round.

haa right, I see what you mean. that shouldn't really happen: websocket do not reorder messages (as it's over TCP). I guess as you added the loop, it got reorder anyway so that might be hinting that the message processing stack is missbehaving.

don't worry to much about it, if you think that's needed, you know best :)

Reply:
Aaah I was assuming we were working with UDP that makes things much easier great, then indeed then any message reorder should indicate something is wrong with the logic rather than the network